compiler/hash-typecheck/src/ops/validate.rs

```diff
@@ -88,110 +298,388 @@ impl<'gs, 'ls, 'cd> Validator<'gs, 'ls, 'cd> {
+                    }
+                    Ok(result)
                 }
-                Level1Term::Fn(_) => {
+                Level1Term::Fn(fn_ty) => {
                     // Validate parameters and return type
-                    todo!()
+                    let fn_ty = fn_ty.clone();
+                    self.validate_params(&fn_ty.params)?;
+                    self.validate_term(fn_ty.return_ty)?;
+
+                    // Ensure each parameter and return type are runtime instantiable:
+                    for param in fn_ty.params.positional() {
+                        self.ensure_term_is_runtime_instantiable(param.ty)?;
+                    }
+                    self.ensure_term_is_runtime_instantiable(fn_ty.return_ty)?;
+
+                    Ok(result)
                 }
                 Level1Term::ModDef(_) | Level1Term::NominalDef(_) => {
-                    // Nothing to do
+                    // Nothing to do, should have already been validated on creation.
+                    Ok(result)
+                }
+            },
+
+            // Level 0 terms:
+            Term::Level0(level0_term) => match level0_term {
+                Level0Term::Rt(rt_inner_term) => {
+                    // Validate the inner term, and ensure it is runtime instantiable:
+                    let rt_inner_term = *rt_inner_term;
+                    self.validate_term(rt_inner_term)?;
+                    self.ensure_term_is_runtime_instantiable(rt_inner_term)?;
+                    Ok(result)
+                }
+                Level0Term::FnLit(fn_lit) => {
+                    // Ensure the inner type is a function type, and get it:
+                    let fn_lit = *fn_lit;
+                    match self.term_is_fn_ty(fn_lit.fn_ty)? {
+                        Some(fn_ty) => {
+                            // Validate constituents:
+                            let fn_ty = fn_ty;
+                            self.validate_params(&fn_ty.params)?;
+                            let fn_return_ty_validation = self.validate_term(fn_ty.return_ty)?;
+                            let fn_return_value_validation =
+                                self.validate_term(fn_lit.return_value)?;
+
+                            // Ensure the return type of the function unifies with the type of the return
+                            // value:
+                            let _ = self.unifier().unify_terms(
+                                fn_return_value_validation.term_ty_id,
+                                fn_return_ty_validation.simplified_term_id,
+                            );
+
+                            // @@Correctness: should we not apply the above substitution somewhere?
+                            Ok(result)
+                        }
+                        // This isn't a user error, it is a compiler error:
+                        None => panic!("Found non-function type in function literal term!"),
```

Comment:
Maybe add a note here about using the upcoming report and panic functionality here

Reply:
It’s difficult to determine where the current traversal is at when it does panic, I thought adding a similar macro to `panic` would be simpler

Reply:
Added general issue for this #306.